revocation/internal/crl/crl.go

```diff
@@ -0,0 +1,298 @@
+			break
+		}
+		crlResults = append(crlResults, crlResult)
+
+		if crlResult.Result == result.ResultRevoked {
+			return &result.CertRevocationResult{
+				Result:     result.ResultRevoked,
+				CRLResults: crlResults,
+			}
+		}
+	}
+
+	if lastErr != nil {
+		crlResults = append(crlResults, &result.CRLResult{
+			Result: result.ResultUnknown,
+			URI:    crlURL,
+			Error:  lastErr,
+		})
+		return &result.CertRevocationResult{
+			Result:     result.ResultUnknown,
+			CRLResults: crlResults,
+		}
+	}
+
+	return &result.CertRevocationResult{
+		Result:     result.ResultOK,
+		CRLResults: crlResults,
+	}
+}
+
+// Supported checks if the certificate supports CRL.
+func Supported(cert *x509.Certificate) bool {
+	return cert != nil && len(cert.CRLDistributionPoints) > 0
+}
+
+func validate(crl *x509.RevocationList, issuer *x509.Certificate) error {
+	// check signature
+	if err := crl.CheckSignatureFrom(issuer); err != nil {
+		return fmt.Errorf("CRL signature verification failed: %w", err)
+	}
+
+	// check validity
+	if !crl.NextUpdate.IsZero() && time.Now().After(crl.NextUpdate) {
+		return fmt.Errorf("CRL is expired: %v", crl.NextUpdate)
+	}
+
+	// unsupported critical extensions is not allowed. (See RFC 5280, Section 5.2)
+	for _, ext := range crl.Extensions {
+		if ext.Critical {
+			return fmt.Errorf("CRL contains unsupported critical extension: %v", ext.Id)
+		}
+	}
+
+	return nil
+}
+
+// checkRevocation checks if the certificate is revoked or not
+func checkRevocation(cert *x509.Certificate, baseCRL *x509.RevocationList, signingTime time.Time, crlURL string) (*result.CRLResult, error) {
+	if cert == nil {
+		return nil, errors.New("certificate is nil")
+	}
+
+	if baseCRL == nil {
+		return nil, errors.New("CRL is nil")
+	}
+
+	// tempRevokedEntries contains revocation entries with reasons such as
+	// CertificateHold or RemoveFromCRL.
+	//
+	// If the certificate is revoked with CertificateHold, it is temporarily
+	// revoked. If the certificate is shown in the CRL with RemoveFromCRL,
+	// it is unrevoked.
+	var tempRevokedEntries []x509.RevocationListEntry
+
+	for _, revocationEntry := range baseCRL.RevokedCertificateEntries {
+		if revocationEntry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+			extensions, err := parseEntryExtensions(revocationEntry)
+			if err != nil {
+				return nil, err
+			}
+
+			// validate signingTime and invalidityDate
+			if !signingTime.IsZero() && !extensions.invalidityDate.IsZero() &&
+				signingTime.Before(extensions.invalidityDate) {
+				// signing time is before the invalidity date which means the
+				// certificate is not revoked at the time of signing.
+				continue
+			}
+
+			if int(result.CRLReasonCodeCertificateHold) == revocationEntry.ReasonCode ||
+				int(result.CRLReasonCodeRemoveFromCRL) == revocationEntry.ReasonCode {
+				// temporarily revoked
+				tempRevokedEntries = append(tempRevokedEntries, revocationEntry)
```

Comment:
Do we need to check `revocationEntry.RevocationTime` against zero time?

Reply:
In fact, we don't care about the revocation time. As long as the revocation entry exists, it is considered revoked. We do check the `invalidityDate`, which can be used to compare with the `signingTime`.